### ISSUE TITLE
Improve audit logging

### ISSUE THREAD

**Issue by emilvirkki** [partio-scout/reki#262]
Currently, REKI audits all user access to and modification of particpant information. However, our audit logs could be even better to improve resilience and to make investigation of incidents easier.

It should be possible to store the audit logs in the database or to export them to a separate logging server.

Log content
---

Our audit logs should include at least the following fields:

- subject (user or null if not applicable)
- timestamp (UTC)
- IP address (could maybe also be part of meta?)
- action (search_participants, create_user etc.)
- objects (the users or participants affected)
- meta (additional information about this action)

For users, we should store the following fields from the time the log entry was created:

- database id
- member number (for identifying accross systems)
- first name and last name (to provide human-readable printouts when an audit log excerpt needs to be produced)

For participants, we should store the following fields:

- database id
- member number, if any (for finding log items affecting a certain person)
- email (for finding log items affecting a certain person who doesn't have a member number)

Actions
---

Actions should include at least:

- view participant
  - meta: what fields were included in the response
- search participants
  - meta: search criteria used
  - meta: what fields were included in the response
- update participant
  - meta: new values
- create user
- lock/unlock user
- delete user
- update user
- login
  - meta: method used (pw / partioid)
- logout

Open questions
---

- Should we include also failed logins and any authorisation issues?
- Is IP necessary in every request, or is it enough to record it for login only? This would make implementation easier. Also, it could be enough to match the IP from the request log of the application using a timestamp.

Some ideas about the implementation
---

Maybe something as simple as:

```javascript
audit(event, subject, object, meta);
```

So, for example, when searching participants:

```javascript
const participants = ...
res.json(participants);
audit('search_participants', req.user, participants, {
  search: req.query.filter,
  ip: req.ip
});
```

How to do the logging so that it can be done either to the database or a centralised logging service needs a bit more thinking.

**Comment by googol** [comment]
Meta for view/list participants: what properties were shown

**Comment by googol** [comment]
Object should probably be just a list of ids, so that we don't get too much PII in our access logs (or too much data in general)

**Comment by emilvirkki** [comment]
The problem with storing just id:s is that if we want to store the audit logs longer than the actual database, the id:s don't provide any information.

**Comment by googol** [comment]
Yeah. I actually had skipped over the part where you listed the saved object data. Sounds reasonable

**Comment by emilvirkki** [comment]
Yeah, I just edited it, sorry! :D

**Comment by googol** [comment]
Closed by @supertassu's PR #266 

Thanks!
